lib/Dialect/TritonGPU/Transforms/RemoveLayoutConversions.cpp

```diff
@@ -82,542 +82,676 @@ class DecomposeDotOperand : public mlir::RewritePattern {
+    Value convertedOperand = operand;
+    if (layouts.count(result))
+      convertedOperand =
+          getValueAs(operand, *layouts[result].encodings.begin());
+    operands.push_back(convertedOperand);
+  }
+  auto newForOp = rewriter.create<scf::ForOp>(
+      forOp.getLoc(), forOp.getLowerBound(), forOp.getUpperBound(),
+      forOp.getStep(), operands);
+
+  newForOp.getBody()->getOperations().splice(
+      newForOp.getBody()->getOperations().begin(),
+      forOp.getBody()->getOperations());
+
+  for (auto [oldResult, newResult] :
+       llvm::zip(forOp.getResults(), newForOp.getResults())) {
+    if (oldResult.getType() == newResult.getType()) {
+      oldResult.replaceAllUsesWith(newResult);
+      continue;
+    }
+    map(oldResult, newResult);
+  }
+
+  for (auto [oldArg, newArg] : llvm::zip(forOp.getBody()->getArguments(),
+                                         newForOp.getBody()->getArguments())) {
+    if (oldArg.getType() == newArg.getType()) {
+      oldArg.replaceAllUsesWith(newArg);
+      continue;
+    }
+    map(oldArg, newArg);
+  }
+  opToDelete.push_back(forOp.getOperation());
+  return newForOp.getOperation();
+}
 
-    IRMapping mapping;
-    rematerializeConversionChain(toConvert, rewriter, processed, mapping);
-    rewriter.replaceOp(cvt, mapping.lookup(cvt->getOperand(0)));
+Operation *LayoutPropagation::rewriteYieldOp(scf::YieldOp yieldOp) {
+  OpBuilder rewriter(yieldOp);
+  Operation *newYield = rewriter.clone(*yieldOp.getOperation());
+  Operation *parentOp = yieldOp->getParentOp();
+  for (OpOperand &operand : yieldOp->getOpOperands()) {
+    Value result = parentOp->getResult(operand.getOperandNumber());
+    auto tensorType = result.getType().dyn_cast<RankedTensorType>();
+    if (!tensorType)
+      continue;
+    Value newOperand = getValueAs(operand.get(), tensorType.getEncoding());
+    newYield->setOperand(operand.getOperandNumber(), newOperand);
+  }
+  opToDelete.push_back(yieldOp.getOperation());
+  return newYield;
+}
 
-    return mlir::success();
+Operation *LayoutPropagation::rewriteOp(Operation *op) {
+  if (auto forOp = dyn_cast<scf::ForOp>(op)) {
+    return rewriteForOp(forOp);
   }
-};
+  OpBuilder rewriter(op);
+  Attribute encoding = *layouts[op->getResult(0)].encodings.begin();
+  if (auto convertOp = dyn_cast<triton::gpu::ConvertLayoutOp>(op)) {
+    Attribute srcEncoding = *layouts[convertOp.getOperand()].encodings.begin();
+    Value src = getValueAs(convertOp.getOperand(), srcEncoding);
+    auto tensorType = op->getResult(0).getType().cast<RankedTensorType>();
+    auto newType = RankedTensorType::get(tensorType.getShape(),
+                                         tensorType.getElementType(), encoding);
+    auto cvt = rewriter.create<triton::gpu::ConvertLayoutOp>(op->getLoc(),
+                                                             newType, src);
+    map(op->getResult(0), cvt.getResult());
+    opToDelete.push_back(op);
```

Comment:
Is it possible to have a more general pattern (one centralized place) that adds ops to `opDelete`?

Reply:
you're right, it can be done at the top of this function. Changing it.